enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,647 @@
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and controllers with the Assisted Service's capabilities.
+
+Hive's ClusterDeployment resource will be extended to serve as the cluster
+definition for agent-based installations. Additional CRDs described below will
+be used by other controllers and actors to perform agent-based provisioning.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For hub clusters that are used for multi-cluster creation and management,
+persistent storage will be a requirement for the database and file system
+storage.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD directly.
+In the meantime, status will be propagated to the appropriate k8s resource when
+an agent reports relevant information.
+
+**InstallEnv**
+This new resource, part of the assisted intaller's new operator, represents an
+environment in which a group of hosts share settings related to networking,
+local services, etc. This resource is used to create a discovery image that
+should be used for booting hosts. In the REST API this corresponds to the
+"image" resource that is embedded in the "cluster" resource, but for
+kubernetes-native it is more natural for it to be separate.
+
+The discovery ISO can be downloaded from a URL that is available in the
+resource's Status.
+
+The details of this resource definition were discussed [in a
+pull request](https://github.com/openshift/assisted-service/pull/969).
+
+Summary of fields:
+
+Spec
+* egress proxy settings
+* a list of NTP sources (hostname or IP) to be added to all cluster
+hosts. They are added to any NTP sources that were configured through other
+means.
+* a list of SSH public keys that will be added to all agents for use in debugging.
+* pull secret reference
+* a label selector for Agents. This is how Agent resources are identified as
```

Comment:
Label selectors can allow a resource to match multiple queries. Do we want to allow an agent to match multiple environments? Would we be better of making the link between the agent and the environment explicit by adding a reference to the agent API pointing to an InstallEnv?

Reply:
An agent should match a single InstallEnv.  @dgoodwin @filanov  I think this is a good suggestion, what do you think?

Reply:
Yes I don't think there's any late binding or choose from pool functionality here, makes sense that the Agent would reference it's InstallEnv explicitly.

Reply:
An explicit reference on the Agent could be useful, but I don't know of a direct use case for it yet. We could add it later potentially, or now if there is a use case.

While the label selector can potentially behave as you describe, the same would apply to a Deployment, MachineSet, or other similar resources.

Typically when you create a Deployment, you make a simple label selector and put the corresponding label in the Pod template a few lines below. The same approach would apply here and should make it simple to avoid conflicts.

Reply:
ITSM that a controller replicating a resource using a template has very different needs from this case. The InstallEnv controller isn't doing anything with agents, and we just want to establish the relationship? And the agent labels you set on InstallEnv are really for the ClusterDeployment controller to identify a set of agents?

Reply:
The InstallEnv controller creates the ISO that is basically a template for Agents. I see it as very similar in that way. The one and only place to define an Agent is the InstallEnv spec. Based on that, utilizing an ISO and some boot process, n number of Agents get stamped out.

Generally if a user or client wants to do a list query with some selector, a label is the way to do that. InstallEnv is a well-defined grouping for Agents and part of their identity, so a label referencing the InstallEnv seems to be the right way to reflect that relationship.  `kubectl get Agents -l 'installenv=foo'` or similar.

We haven't dug into reporting details in this proposal, but there is value in reporting on an InstallEnv's aggregate Agent status, including how many Agents were booted from an out-of-date ISO for example. We could imagine future actions that may be taken on an Agent based on changes to its InstallEnv, though that's not in-scope for this enhancement.

Another related detail not mentioned but that's probably implied is that an Agent should likely have an owner reference to its InstallEnv. Agent deletion in general may still have some open questions to work out, but I think it will make sense to garbage-collect them if their InstallEnv gets deleted. In any case, the owner reference may serve the role of the explicit reference you're looking for.

Reply:
> The InstallEnv controller creates the ISO that is basically a template for Agents. I see it as very similar in that way. The one and only place to define an Agent is the InstallEnv spec. Based on that, utilizing an ISO and some boot process, n number of Agents get stamped out.

I think it's unlikely users will make that mental connection with the current design.

How would you feel about taking the next logical step with that analogy - putting the labels under a AgentTemplateSpec field, like installEnv.spec.template.metadata.labels ? Could we imagine wanting to extend AgentTemplateSpec as we want to allow the user to specify additional stuff that would show up in the created Agent resources?

> Generally if a user or client wants to do a list query with some selector, a label is the way to do that. InstallEnv is a well-defined grouping for Agents and part of their identity, so a label referencing the InstallEnv seems to be the right way to reflect that relationship. `kubectl get Agents -l 'installenv=foo'` or similar.
> 
> We haven't dug into reporting details in this proposal, but there is value in reporting on an InstallEnv's aggregate Agent status, including how many Agents were booted from an out-of-date ISO for example. We could imagine future actions that may be taken on an Agent based on changes to its InstallEnv, though that's not in-scope for this enhancement.

Both of those seem compelling to me, thanks

Reply:
I like that thinking, but I'm not able to come up with anything other than labels that we would want to templatize for the Agent. (Also note that labels are not part of the Agent's Spec, but the Metadata.) Most of what the InstallEnv is providing as input to Agent creation is data that's required to run the live ISO, which of course precedes the Agent resource. Maybe there's a use case for adding custom annotations that are meaningful to the user, in which case we could templatize the Agent's metadata. Imagine the [ReplicaSet template](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#example) without the `spec` part. But the annotation use case may or may not be terribly useful.

Right now, this is what the labels field looks like on the InstallEnv Spec:

```
	// AgentLabels lists labels to apply to Agents that are associated with this InstallEnv upon
	// the creation of the Agents.
	AgentLabels map[string]string `json:"agentLabels,omitempty"`
```

Another option is we could try to name this field more clearly. WDYT?

Reply:
I like the metadata template approach, if you think that can work well